lib/spack/spack/package.py

```diff
@@ -66,6 +66,40 @@
 """Allowed URL schemes for spack packages."""
 _ALLOWED_URL_SCHEMES = ["http", "https", "ftp", "file", "git"]
 
+import threading
+
+class DoInstallThread(threading.Thread):
+  """
+  This is a lightweight thread class to take care of executing
+  do_install calls in parallel.
+
+  To avoid the same package to be build multiple times,
+  the queued builds are being recorded. If a package thread finds
+  its target package already being in the queue, it just waits for
+  the other thread to finish.
+
+  """
+  _do_install_threads = {}
+  _queue_lock = threading.Lock()
+
+  def __init__(self, package, kwargs):
```

Comment:
@hegner: do you find that this gets you significant speed up on-node?  I think the big win here would be that multiple fetches could proceed in parallel, which is something spack doesn't do normally.  I'm not so sure that actually building in parallel gets you much in most cases, as most builds can use the whole node already.  I guess I'm curious whether we want to still limit to one build at a time on node but fetch in parallel, and put tarballs that arrive on a ready queue...